src/catalog/database_catalog.cpp

```diff
@@ -1774,6 +1784,33 @@ Column DatabaseCatalog::MakeColumn(storage::ProjectedRow *const pr, const storag
   return col;
 }
 
+bool DatabaseCatalog::TryLock(transaction::TransactionContext *const txn) {
+  auto current_val = write_lock_.load();
+
+  const transaction::timestamp_t txn_id = txn->FinishTime();     // this is the uncommitted txn id
+  const transaction::timestamp_t start_time = txn->StartTime();  // this is the unchanging start time of the txn
+
+  const bool already_hold_lock = current_val == txn_id;
+  if (already_hold_lock) return true;
+
+  const bool owned_by_other_txn = (!transaction::TransactionUtil::Committed(current_val) && !already_hold_lock);
+  const bool newer_committed_version = transaction::TransactionUtil::Committed(current_val) &&
+                                       transaction::TransactionUtil::NewerThan(current_val, start_time);
+
+  if (owned_by_other_txn || newer_committed_version) return false;
```

Comment:
Do we not have a function/macro that encapsulates this logic?  It seems like we should since this is identical to asking "do I have a write-write conflict".

Reply:
It's HasConflict in the storage layer but the function signature is different. I can see what I can come up with.